pkg/virt-handler/vm.go

```diff
@@ -500,46 +528,67 @@ func (d *VirtualMachineController) getLauncherClient(vmi *v1.VirtualMachineInsta
 
 func (d *VirtualMachineController) processVmShutdown(vmi *v1.VirtualMachineInstance, domain *api.Domain) error {
 
-	clientDisconnected := false
+	// Only attempt to shutdown/destroy if we still have a connection established with the pod.
+	client, err := d.getVerifiedLauncherClient(vmi)
 
-	client, err := d.getLauncherClient(vmi)
-	if err != nil {
-		clientDisconnected = true
-	}
+	// If the pod has been torn down, we know the VirtualMachineInstance is down.
+	if err == nil {
 
-	// verify connectivity before processing shutdown.
-	// It's possible the pod has already been torn down along with the VirtualMachineInstance.
-	if clientDisconnected == false {
-		err := client.Ping()
-		if cmdclient.IsDisconnected(err) {
-			clientDisconnected = true
-		} else if err != nil {
+		// Only attempt to gracefully shutdown if the domain has the ACPI feature enabled
+		if useShutdown(vmi, domain) {
+			expired, timeLeft := d.hasGracePeriodExpired(domain)
+			if !expired && domain.Status.Status != api.Shutdown {
+				err = client.ShutdownVirtualMachine(vmi)
+				if err != nil && !cmdclient.IsDisconnected(err) {
+					// Only report err if it wasn't the result of a disconnect.
+					return err
+				}
+
+				log.Log.Object(vmi).Infof("Signaled graceful shutdown for %s", vmi.GetObjectMeta().GetName())
+				// pending graceful shutdown.
+				d.Queue.AddAfter(controller.VirtualMachineKey(vmi), time.Duration(timeLeft)*time.Second)
+				d.recorder.Event(vmi, k8sv1.EventTypeNormal, v1.ShuttingDown.String(), "Signaled Graceful Shutdown")
+				return nil
+			} else if !expired {
```

Comment:
That  `!expired` means already shut down is very hard to understand.

Reply:
Agree. Will make 2 nested ifs, should make it clearer.